### PR TITLE
add dnfmodule repo implementation

### PR DESCRIPTION
**Pull request by evgeni**

### CODE REVIEW

File: manifests/repo/dnfmodule.pp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#
+# @api private
+class nodejs::repo::dnfmodule (
+  String[1] $ensure = $nodejs::repo_version,
+  String[1] $module = 'nodejs',
+) {
+  package { 'nodejs dnf module':
+    ensure      => $ensure,
+    name        => $module,
+    enable_only => true,
```
Comment (ekohl) [inline manifests/repo/dnfmodule.pp:21]: I'm not sure if this is still needed. I've dropped it in other places so solve upgrading issues.

Comment (evgeni) [reply manifests/repo/dnfmodule.pp]: not where I've copied it from XD

https://github.com/voxpupuli/puppet-redis/blob/ee9956b10586284f3cbecc26baa2b79f35da0fc5/manifests/dnfmodule.pp#L21

File: manifests/repo/dnfmodule.pp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+#
+# @param module
+#   Name of the nodejs dnf package
+#
+# @api private
+class nodejs::repo::dnfmodule (
+  String[1] $ensure = $nodejs::repo_version,
+  String[1] $module = 'nodejs',
+) {
+  package { 'nodejs dnf module':
```
Comment (evgeni) [inline manifests/repo/dnfmodule.pp:18]: Should we add a `fail()` for non RH family or major ≠ 8/9?

Comment (ekohl) [reply manifests/repo/dnfmodule.pp:18]: Looking at `nodejs::repo::nodesource` it'd be consistent, though technically it also works on Fedora. At least Fedora 38 still has a NodeJS 16 module (while 18 non-modular is the default).

Given people who look for modularity know what it is, I wouldn't be too strict about it.

**Review by ekohl — APPROVED** [review]

**Comment by ekohl** [comment]
This is a new one for me:

```
        	broken groups or modules: nodejs:18
       	Error: /Stage[main]/Nodejs::Repo::Dnfmodule/Package[nodejs dnf module]/ensure: change from 'purged' to '18' failed: Could not update: Execution of '/usr/bin/dnf module install -d 0 -e 1 -y nodejs:18' returned 1: Error: Problems in request:
       	broken groups or modules: nodejs:18
```

**Comment by evgeni** [comment]
It started once I dropped the enable_only 🤔

**Comment by ekohl** [comment]
It made me look it up: `enable_only` makes it `dnf module enable` instead of `dnf module install` (https://github.com/puppetlabs/puppet/blob/5dae3b1f808d4fe0def1c5673c65273d2466ff43/lib/puppet/provider/package/dnfmodule.rb#L88-L103). I guess that doesn't work for the `nodejs` module.

**Comment by ekohl** [comment]
And there is code that should catch that error, but looks like that doesn't work.

File: spec/acceptance/class_spec.rb [hunk 3]
```diff
@@ -65,6 +65,42 @@ class { 'nodejs':
...
+      npm
+      nodejs
+      nodejs-devel
+    ].each do |pkg|
+      describe package(pkg) do
+        it do
+          is_expected.to be_installed
+        end
+
+        it 'comes from the expected source' do
```
Comment (evgeni) [inline spec/acceptance/class_spec.rb:96]: Add a `it 'has the right version'` that asserts the output of `shell('node --version')` to be match `v18` (or well, whatever we install)

**Comment by evgeni** [comment]
Modules really, really, should never be installed (it was a mistake to allow that). (end-of-rant)

**Comment by evgeni** [comment]
> And there is code that should catch that error, but looks like that doesn't work.

The match is: `/^(?:missing|broken) groups or modules: #{Regexp.quote(@resource[:name])}$/`, so in our case: `^broken groups or modules: nodejs$`, but the output is "… nodejs:18".

**Comment by evgeni** [comment]
also, no idea why exactly 18 has no default profile:

```
# dnf module list nodejs
Failed to set locale, defaulting to C.UTF-8
Last metadata expiration check: 0:07:22 ago on Sun Oct 29 17:42:59 2023.
CentOS Stream 8 - AppStream
Name                                            Stream                                         Profiles                                                                      Summary                                                   
nodejs                                          10 [d]                                         common [d], development, minimal, s2i                                         Javascript runtime                                        
nodejs                                          12                                             common [d], development, minimal, s2i                                         Javascript runtime                                        
nodejs                                          14                                             common [d], development, minimal, s2i                                         Javascript runtime                                        
nodejs                                          16                                             common [d], development, minimal, s2i                                         Javascript runtime                                        
nodejs                                          18                                             common, development, minimal, s2i                                             Javascript runtime                                        
nodejs                                          20                                             common [d], development, minimal, s2i                                         Javascript runtime                                        

Hint: [d]efault, [e]nabled, [x]disabled, [i]nstalled
```

**Comment by evgeni** [comment]
> also, no idea why exactly 18 has no default profile

LOOK MA, TWO BUGS FOR THE PRICE OF ONE!

```console
# dnf module list nodejs
Red Hat Enterprise Linux 8 for x86_64 - AppStream (RPMs)
Name                                           Stream                                            Profiles                                                                     Summary                                                  
nodejs                                         10 [d][e]                                         common [d], development, minimal, s2i                                        Javascript runtime                                       
nodejs                                         12                                                common [d], development, minimal, s2i                                        Javascript runtime                                       
nodejs                                         14                                                common [d], development, minimal, s2i                                        Javascript runtime                                       
nodejs                                         16                                                common [d], development, minimal, s2i                                        Javascript runtime                                       
nodejs                                         18                                                common [d], development, minimal, s2i                                        Javascript runtime                                       
```

**Review by ekohl — APPROVED** [review]

**Comment by evgeni** [comment]
> LOOK MA, TWO BUGS FOR THE PRICE OF ONE!

https://issues.redhat.com/browse/RHEL-14970

**Comment by evgeni** [comment]
> I think the match should be `/^(?:missing|broken) groups or modules: #{Regexp.quote(args)}$/` instead

https://github.com/puppetlabs/puppet/pull/9138

**Pull request merged**